src/mono/mono/metadata/assembly.c

```diff
@@ -3182,10 +3181,21 @@ mono_create_new_bundled_satellite_assembly (const char *name, const char *cultur
 /**
  * mono_register_bundled_satellite_assemblies:
  */
+MONO_DEPRECATED
 void
-mono_register_bundled_satellite_assemblies (const MonoBundledSatelliteAssembly **assemblies)
-{
-	satellite_bundles = assemblies;
+mono_register_bundled_satellite_assemblies (const MonoBundledSatelliteAssembly **satellite_assemblies)
+{
+	for (int i = 0; satellite_assemblies [i]; ++i) {
+		const MonoBundledSatelliteAssembly *satellite_assembly = satellite_assemblies [i];
+		char *id = g_strconcat (satellite_assembly->culture, "/", satellite_assembly->name, (const char*)NULL);
+		MonoBundledSatelliteAssemblyResource *satellite_assembly_resource = mono_bundled_resources_get_satellite_assembly_resource (id);
+		g_assert (!satellite_assembly_resource);
+		satellite_assembly_resource->resource.type = MONO_BUNDLED_SATELLITE_ASSEMBLY;
+		satellite_assembly_resource->resource.id = id;
+		satellite_assembly_resource->resource.free_bundled_resource_func = &mono_bundled_resources_free_bundled_resource_func;
+		satellite_assembly_resource->satellite_assembly = *(MonoBundledSatelliteAssemblyData *)satellite_assembly;
```

Comment:
As in at some point we might make `MonoBundled(Satellite)Assembly` and `MonoBundled(Satellite)AssemblyData` differ in struct layout? Or more that the types are different as one is mono internal types and the other is using C99 compliant types?

Reply:
Since you cast stuff from one struct type to another that are potentially "unrelated", there is no way for compiler to catch if the layout changes on either side making them incompatible. Compiler will mainly just copy the size number of bytes of MonoBundledSatelliteAssemblyData from satellite_assembly to satellite_assembly_resource->satellite_assembly, without any knowledge on what underlying data actually represents. If you on the other hand just extracted the known fields and placed them into corresponding known fields then compiler will also be able to type check the individual assignment for incompatible data types.